core/src/main/java/hudson/cli/DeleteCommand.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2015 Red Hat, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package hudson.cli;
+
+import hudson.AbortException;
+import jenkins.model.Jenkins;
+
+import java.util.List;
+import java.util.HashSet;
+
+/**
+ * CLI command, which helps delete job(s) or node(s).
+ * @since TODO
+ */
+
+public abstract class DeleteCommand<T> extends CLICommand {
```

Comment:
Probably better to start out `@Restricted`.

Reply:
I assume `@Restricted` means restricting the type parameter? I'm not familiar with a good way of doing this without polluting the type classes with a wrapper interface.

Reply:
`@Restricted(NoExternalUse.class)` means it's not available to plugins as API. Since plugins are often written against older releases of Jenkins, it's difficult to maintain binary compatibility if they use whatever internal classes they like just because they're `public`. So anything that isn't intended to be used by plugins should just be `@Restricted` to prevent even accidental use in plugins.

Think of it as class and method level module exports, but pre-dating Java 9.

Reply:
Understood, thanks for the explanation.

Reply:
Line comments are for the line above the comment, not the line below the comment. Annotate the class.